lib/ohai/plugins/ec2.rb

```diff
@@ -122,10 +122,15 @@ def looks_like_ec2?
       fetch_metadata.each do |k, v|
         # fetch_metadata returns IAM security credentials, including the IAM user's
         # secret access key. We'd rather not have ohai send this information
-        # to the server.
-        # http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/AESDG-chapter-instancedata.html#instancedata-data-categories
-        next if k == "iam" && !hint?("iam")
-
+        # to the server. If the instance is associated with an IAM role we still
+        # grab the name of the role only.
+        # https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instancedata-data-categories.html
+        if k == "iam" && !hint?("iam")
+          if v["security-credentials"] && v["security-credentials"].keys.length == 1
+            ec2[:role_name] = v["security-credentials"].keys[0]
+          end
+          next
+        end
         ec2[k] = v
```

Comment:
```suggestion
        if k == "iam" && !hint?("iam")
          if v["security-credentials"] && v["security-credentials"].keys.length == 1
            ec2[:role_name] = v["security-credentials"].keys[0]
          end
        else
          ec2[k] = v
        end
```

The control flow could be simplified here.